### PR TITLE
Update GIMP plugin to use public API and add save dialog

### PR DESCRIPTION
**Pull request by xiota**
* Remove libjxl internals
* Use multi-threaded parallel runner
* Allow saving without requiring compiler flag
* Load multiple frames as layers
* Save non-linear uint8/16 and linear float16/32
* Save alpha channel in uint8/16 formats
* Add save options dialog

The following issues are resolved: #100, #229, #417, [wg1#234](https://gitlab.com/wg1/jpeg-xl/-/issues/234), [wg1#184](https://gitlab.com/wg1/jpeg-xl/-/issues/184).

save_dialog

### CODE REVIEW

File: plugins/gimp/file-jxl-load.cc
Comment (sboukortt) [inline plugins/gimp/file-jxl-load.cc:25]: This is rather minor but could we use `nullptr`, and rename `loadlinear` to `load_linear`?

File: plugins/gimp/file-jxl-load.cc
Comment (sboukortt) [inline plugins/gimp/file-jxl-load.cc:180]: I think this will reintroduce https://gitlab.com/wg1/jpeg-xl/-/issues/182. Ideally, we would also try to get the color profile as a `JxlColorEncoding` and, if successful, compare it against sRGB and use `gimp_color_profile_new_rgb_srgb` if it matches. Would you like to give it a try in this pull request or should we do it later?

File: plugins/gimp/file-jxl-save.cc
Comment (sboukortt) [inline plugins/gimp/file-jxl-save.cc:16]: This seems to be a leftover.

File: plugins/gimp/file-jxl-save.cc
Comment (sboukortt) [inline plugins/gimp/file-jxl-save.cc:34]: In C++, you can write this simply as:

```suggestion
struct JpegXlSaveVals {
  gboolean tmp;
};
```

File: plugins/gimp/file-jxl-save.cc
Comment (sboukortt) [inline plugins/gimp/file-jxl-save.cc:40]: Likewise, `(void)` is a remnant from C:

```suggestion
bool save_dialog() {
```

File: plugins/gimp/file-jxl-save.cc
Comment (sboukortt) [inline plugins/gimp/file-jxl-save.cc:306]: I feel that I must be missing something but shouldn’t we get the profile from GIMP?

**Review by sboukortt — COMMENTED** [review]
Thank you, it’s great to have this!

File: plugins/gimp/file-jxl-load.cc
Comment (xiota) [reply plugins/gimp/file-jxl-load.cc:25]: No problem.  I'll make the change and update the commit.

File: plugins/gimp/file-jxl-load.cc
Comment (xiota) [reply plugins/gimp/file-jxl-load.cc:180]: I didn't notice the regression because I had GIMP set to keep embedded profiles.  The problem is Thread Pool Internal has access to CodecInOut, but Parallel Runner doesn't, as far as I can tell.  I'll take a look at it.

File: plugins/gimp/file-jxl-save.cc
Comment (xiota) [reply plugins/gimp/file-jxl-save.cc:34]: Thanks.  I'll change this.  It's been a long time since I've done anything with a C-like language, and all of the plugins in the GIMP repository are written in C...

File: plugins/gimp/file-jxl-save.cc
Comment (xiota) [reply plugins/gimp/file-jxl-save.cc:16]: `glib/gi18n.h` is used for the `_(string)` wrappers.  Not sure whether to keep them or not.

I also don't know whether to put `libgimp/gimp*` in `file-jxl-save.h`.

`glib/gstdio.h` is commented out because I didn't know whether I'd need it later.  So far, it looks like I won't, so I'll probably just remove it.

File: plugins/gimp/file-jxl-save.cc
Comment (xiota) [reply plugins/gimp/file-jxl-save.cc:306]: Yes, getting the profile from GIMP would make sense.  I either overlooked the old code that did that or didn't know how to do it.  I'll take a look at it and update the commit if I can figure it out.

Comment (xiota) [reply plugins/gimp/file-jxl-save.cc:306]: So my left hand doesn't know what my right hand is doing... there's some icc processing at the beginning of the file, but it doesn't go anywhere... Also looks like another CodecInOut access difference between Thread Pool Internal and Parallel Runner.  I'll poke around some more to see if there's something I overlooked earlier..

File: plugins/gimp/file-jxl-load.cc
Comment (xiota) [reply plugins/gimp/file-jxl-load.cc]: I think the regression is fixed now...  it's a bit complicated because I'm avoiding using internal headers.  (I'd like to remove "lib/jxl/base/file_io.h" someday.)

File: plugins/gimp/file-jxl-save.cc
Comment (xiota) [reply plugins/gimp/file-jxl-save.cc]: ICC profiles should be working more or less as expected now.  If internal GIMP profiles are used in the image, then internal JPEG XL profiles will be embedded in the image.  Otherwise the user-specified ICC profile is embedded.

There are also some limits on when linear or perceptual gamma can be used.  Linear only works with float and perceptual only with int because images with linear int or perceptual float don't display properly.  Conversion by GIMP without any user intervention.

**Comment by xiota** [comment]
@sboukortt I made the changes you suggested.  ICC profiles should work more or less as expected.  Linear/perceptual are converted without user intervention depending on the bit format because certain combinations don't display properly.  The limit can be removed later if the cause is discovered.

I'm also curious about your opinion about whether to set `jxl_info.uses_original_profile` to true or false.  When set to true, only one color profile is saved in the jxl file.  When false, there is an original profile and a data profile.  Files generated by `cjxl` usually have the same profile set for original and data.  However, when saving with the modified GIMP plugin, it's possible to have an original AdobeRGB profile (for instance), and a data sRGB profile.  Then I don't know whether the image should be loaded with the original or data profile.  (I don't know what color transformations the JXL encoder is doing, if any.)

Right now, the modified plugin is saving both profiles and loading using the data profile.  It's a simple change if only one profile should be saved or if images should be loaded using the original profile.

**Comment by novomesk** [comment]
It is a good idea to rewrite the plug-in to use public API.

In order to increase chances for this plug-in to be included by official GIMP, I suggest to try following GIMP coding style: https://gitlab.gnome.org/GNOME/gimp/-/blob/master/CODING_STYLE.md

From GIMP's perspective it would be preferred to use only C instead of C++, that's just FYI.

I have alternative implementation for GIMP 2.99 (it has different API than 2.10.x) in C only. It is not finished but if you want to take a look: https://github.com/novomesk/gimp-jpegxl-plugin/blob/main/src/file-jpegxl.c

**Comment by deymo** [comment]
@novomesk do you plan to upstream your plugin to GIMP repo?

**Comment by xiota** [comment]
@novomesk Converting to C and following GIMP plugin conventions would be a major undertaking that I'll leave for someone else.  I expect GIMP won't have an official plugin until after the JPEG XL standard is published.

Even if I wanted to follow GIMP coding style, I couldn't.  To push to this repository requires passing an automated format check.  It's nice not having to type spaces everywhere, then just letting `clang-format` fix it.

**Comment by novomesk** [comment]
> @novomesk do you plan to upstream your plugin to GIMP repo?

@deymo Yes. GIMP developer knows about the plug-ins and waiting for MR when everything is OK according the expectations.

When contributing to GIMP, it has to be for 2.99.x first and later the new features can be backported to old 2.10.x

In my plug-in I always export in 16bit depth with non-linear sRGB. I chose this for compatibility reasons because lot of viewers do not have color management yet.

**Comment by veluca93** [comment]
> > @novomesk do you plan to upstream your plugin to GIMP repo?
> 
> @deymo Yes. GIMP developer knows about the plug-ins and waiting for MR when everything is OK according the expectations.
> 
> When contributing to GIMP, it has to be for 2.99.x first and later the new features can be backported to old 2.10.x
> 
> In my plug-in I always export in 16bit depth with non-linear sRGB. I chose this for compatibility reasons because lot of viewers do not have color management yet.

I don't think that will work for non-SDR content, among other things... I think it would be a better idea to export things in their current colorspace, and leave it to users to do compatibility decisions :)

Anyway, if you plan to upstream your plugin, I don't think upstreaming this plugin is necessary, no?

**Comment by deymo** [comment]
> Anyway, if you plan to upstream your plugin, I don't think upstreaming this plugin is necessary, no?

Yeah, it looks like @novomesk plugin is already ported to the external API. We could delete this one and add a reference to that one in the README. WDYT?

It would be good to update it to support HDR properly in gimp.

**Comment by deymo** [comment]
Actually, let's keep this plugin around (in C++, with libjxl code style) for using it in 2.10.

**Comment by xiota** [comment]
@deymo  Nooooo.... don't delete this plugin...  Aside from the work I just put into it, considering the historical pace of GIMP development, the new API may not be considered stable for the next decade.  Also, this plugin is able to export to grayscale, 8- and 16-big non-linear, and 32-bit linear formats.

**Comment by novomesk** [comment]
Keep this plug-in for GIMP 2.10, don't delete.

And I will add option for 2.99 plug-in to use original profile during export (otherwise compatible sRGB will be used).

**Comment by xiota** [comment]
@novomesk 
In GIMP 2.10, trying to load 16-bit linear jxl files with the public API doesn't seem to work correctly.  Maybe 2.99 handles it better.

This plugin tries to figure out what profile GIMP is using to set appropriate flags for libjxl (and vice versa).  Otherwise, you can just embed the ICC profile GIMP provides, but that seems to result in disaster since the API doesn't seem to extract the settings from ICC profiles.  When `uses_original_profile=false`, it looks like the public API decodes XYB to sRGB no matter what.  However, if `uses_original_profile=true`, I noticed file sizes double.

**Comment by deymo** [comment]
> When `uses_original_profile=false`, it looks like the public API decodes XYB to sRGB no matter what.

@xiota that's correct. Unlike many other formats, in JXL it is possible to have images encoded in absolute color space like XYB. In other formats like png you provide an array of integers to the encoder, and on the other side you when you decode you get the same integers (or ~similar integers in lossy ones). There's a blob somewhere that tells you what those integers mean in terms of colors (like the ICC blob), but the codec doesn't encode "colors", it encodes numbers.

In the `uses_original_profile=false` case JXL decode API tells you the color encoding of the pixels it returns (normally sRGB) and it tells your the color encoding of the original image as an FYI. The right thing to do depends what you intend to do with the decoded pixels. For example, let's say the original image was 8-bit and had some specific ICC attached (which is not sRGB) but in JXL the image is encoded in XYB, then:
 * if you want to obtain the original 8-bit pixel values you need to decode to float in the jxl api, convert those floats from the jxl-provided (`JXL_COLOR_PROFILE_TARGET_DATA`) color encoding (likely sRGB in this case) to the ICC attached in the file and then reduce the precision to 8-bit integers (you'd likely do this on the output callback to save memory). If you instead get 8-bit integers in sRGB from the API you would incur into clamping and rounding issues.
 * OTOH, if you want to *display* the pixels on a monitor then the original color encoding is basically irrelevant because you'd be converting that image to the color encoding that your output buffer for the display needs, so doing a color conversion directly to the display is likely faster.

I think that for GIMP you always want the first case. If `uses_original_profile=false` then you want to request FLOAT and do the cms conversion in the plugin (the decoder api doesn't use a CMS), in the =true case the pixel data is already in that profile so no conversion is needed in the example above you could directly request 8-bit integers from the API and skip the conversion altogether.

@sboukortt @veluca93 WDYT?

**Comment by xiota** [comment]
> I think that for GIMP you always want the first case. If uses_original_profile=false then you want to request FLOAT and do the cms conversion in the plugin (the decoder api doesn't use a CMS), in the =true case the pixel data is already in that profile so no conversion is needed in the example above you could directly request 8-bit integers from the API and skip the conversion altogether.

@deymo That makes sense, and it seems to be what people would expect from existing image formats and software.  I think I'd like to leave it for a future pull request though.

File: plugins/gimp/file-jxl-save.h
Comment (sboukortt) [inline plugins/gimp/file-jxl-save.h:15]: Given that this and `JpegXlSaveOpts` don’t seem to be used outside of the `.cc` file, it would perhaps be preferable not to declare them in the header file, and to put them in an anonymous namespace to give them internal linkage. (The local naming convention would also be to name the function `SaveDialog`.)

File: plugins/gimp/file-jxl-save.cc
Comment (sboukortt) [inline plugins/gimp/file-jxl-save.cc:578]: FYI, in modern C++, it is also possible to initialize members in the class declaration directly. (But up to you.)

https://en.cppreference.com/w/cpp/language/data_members#Member_initialization (2)

File: plugins/gimp/file-jxl-save.cc
Comment (sboukortt) [inline plugins/gimp/file-jxl-save.cc:26]: C++-style cast would be preferred:

```suggestion
      ("Saving JPEG XL file:" + std::string(filename)).c_str());
```

**Review by sboukortt — APPROVED** [review]
Thanks again!

File: plugins/gimp/file-jxl-save.h
Comment (veluca93) [inline plugins/gimp/file-jxl-save.h:23]: Which one ends up getting used? `distance` or `quality`? I see that in the dialog you have both...

File: plugins/gimp/file-jxl-save.cc
Comment (veluca93) [inline plugins/gimp/file-jxl-save.cc:460]: Just from the option name, I'd consider "effort/speed" somewhat confusing and I'd just leave Effort there. WDYT?

File: plugins/gimp/file-jxl-save.h
Comment (veluca93) [reply plugins/gimp/file-jxl-save.h:23]: Oh, nvm me, apparently they are tied together :)

File: plugins/gimp/file-jxl-save.cc
Comment (xiota) [reply plugins/gimp/file-jxl-save.cc:578]: So this shows how long it's been since I've done anything with C++...  I'd personally rather initialize them directly.  Not sure if I should do that for this commit or at a later time, since copy/paste can lead to errors.

File: plugins/gimp/file-jxl-save.cc
Comment (xiota) [reply plugins/gimp/file-jxl-save.cc:460]: I was planning to change it to just "Encoding Effort", as you suggest.  But I thought all the force pushes may have been delaying review, so I was holding off on it.

File: plugins/gimp/file-jxl-save.cc
Comment (xiota) [reply plugins/gimp/file-jxl-save.cc:26]: Thanks.  Will fix this.

File: plugins/gimp/file-jxl-save.cc
Comment (xiota) [reply plugins/gimp/file-jxl-save.cc:7]: Yes.  Will remove.

File: plugins/gimp/common.cc
Comment (xiota) [reply plugins/gimp/common.cc:27]: Will remove.  Thanks.

File: plugins/gimp/file-jxl-save.h
Comment (xiota) [reply plugins/gimp/file-jxl-save.h:15]: What about the naming of the other functions, like the gnome callback functions?  Would those become GuiOnChangeSuchAndSuch()?

File: plugins/gimp/file-jxl-save.cc
Comment (veluca93) [reply plugins/gimp/file-jxl-save.cc:578]: I think what Sami meant is that you can do 

```
JpegXlSaveOpts::JpegXlSaveOpts(): advanced_mode(false),  lossless(false), ... {
  set_distance(1.0f);
}
```

which I personally would find more readable - but I'm not sure and I don't really have that strong of an opinion :P

File: plugins/gimp/common.h
Comment (xiota) [reply plugins/gimp/common.h:33]: Will change.  Thanks.

File: plugins/gimp/file-jxl-save.cc
Comment (sboukortt) [reply plugins/gimp/file-jxl-save.cc:578]: I was thinking of even putting them outside of the constructor, directly in the class declaration:

```c++
class JpegXlSaveOpts {
 public:
  bool advanced_mode = false;
  bool lossless = false;
  // …
};
```

**Review by xiota — COMMENTED** [review]
@mo271 @sboukortt @veluca93  I've pushed what I hope will be the final changes for this PR.  I made the changes suggested earlier.  Fixed some typos.  Added float16 image saving, and added loading multiple frames as layers.  Thank you for your help.

File: plugins/gimp/common.cc
Comment (sboukortt) [reply plugins/gimp/common.cc:19]: I mean the return statement, since the end of the function is reached anyway.

Comment (xiota) [reply plugins/gimp/common.cc:19]: O, haha.  I try to explicitly return functions.  My editor highlights them so it helps me identify where the end is.

File: plugins/gimp/file-jxl-save.h
Comment (xiota) [reply plugins/gimp/file-jxl-save.h]: My favorite part of the plugin is those sliders.

**Review by sboukortt — APPROVED** [review]

**Review by deymo — APPROVED** [review]
I think there some cleanup to do here but we can do it later.

**Comment by xiota** [comment]
@deymo I know the function names don't follow the local style and should probably be changed.  Is there any other general cleanup you had in mind?

**Comment by jonsneyers** [comment]
Merging now, code style cleanup is not urgent imo for 'external' code like this plugin.

File: plugins/gimp/file-jxl-save.cc
Comment (deymo) [inline plugins/gimp/file-jxl-save.cc:765]: `gimp_color_profile_is_gray` is using the address of the function as a boolean, instead of calling it. I'll send a fix for this.

Comment (xiota) [reply plugins/gimp/file-jxl-save.cc:765]: Thanks.  I saw that after it had already been merged.  It was supposed to be `!jxl_save_opts.is_gray`.

**Pull request merged**
